### DISCUSSION TITLE
reCalculate

### DISCUSSION THREAD

**Discussion by Makhmut2445** [totumonline/totum-ru-issues-and-discussions#146] [Как это сделать?] [answered]
Вопрос касаемо темы  reCalculate.

У меня есть расчётные в цикле таблицы, из них вытягиваются данные и собираются сводные. Когда меняю внутри, данные сразу подтягиваются в эти сводные. Но из этих сводных собираются другие сводные. И в сводных которые составлены из сводных уже не меняются данные которые были изменены внутри, приходится открывать и закрывать все таблицы последовательно.
Как сделать так, чтоб при изменении внутренней таблицы,обновлялись все зависимые от неё таблицы...  При любом изменении. 
Я попытался в рекалкулейт вписать все таблицы в Код действия таблицы, не среагировало...

**Comment by ttm-alexeypolunin** [comment, marked as answer]
В системе `Источник` — `Агрегат` — `Отчет` вам действительно нудно пересчитывать `Агрегат` для того, что бы получить свежие данные в `Отчет`.

Судя по описанию в качестве `Агрегата` у вас `Расчетные в дереве`, раз они пересчитываются при открытии.

Если `Отчет` тоже `Временная` или `Расчетная в дереве` то она будет вычисляться в момент открытия.

И да, вы можете из кода действия таблицы `Источник` пересчитать `Агрегаты`:

```
a1=: recalculate(table: 'table_name_1')

a2=: recalculate(table: 'table_name_2')
```

Если у вас `Агрегат` не `Расчетная таблица`, то надо указать какие строки пересчитывать через параметр `where:`.

Но я бы сказал, что это не самый хороший подход, тк это замедляет взаимодействие с таблицей `Источника`. Я бы сделал обновление `Агрегата` по расписанию, или бы сделал таблицу в которую записывал необходимость обновить `Агрегат` и обновлял бы его при открытии `Отчета`. 

Например если в отчеты вы формируете `insert` в коде, то перед ним при помощи `while` можно выполнить `recalculate` тк эта одна из функций действия, которая может выполняться из кодов. но важно сделать это с проверкой необходимости пересчета.

![Диаграмма без названия drawio](https://user-images.githubusercontent.com/55755565/171159377-a0280f82-7456-43df-90e4-d689f6103755.png)

Код в Report например:

```
=: while(condition: @control.control_field = true; action: $recalc1; action: recalc2; action: $selectData)

recalc1: recalculate(table: 'table_1')

recalc1: recalculate(table: 'table_2')

selectData: ...
```

Небольшая сложность здесь в том, что из кода нельзя выполнить `set` для того, что бы отключить `control`. Для этого нужно сделать поле чекбокс в хедере, которое при любом изменении будет становиться в противоположное значение и по коду-действия этого поля ставить `control` в `false`. По очереди исполнения оно должно быть за полем, которое выполняет `select`.

Чекбокс в хедере Код:

```
= : if(condition: $#onfv = true; then: false; else: true)
```

**Reply by Makhmut2445** [reply]
a1=: recalculate(table: 'table_name_1')
Этот код записывается в Код действия таблицы?

**Reply by ttm-alexeypolunin** [reply]
угу

**Reply by juramebel** [reply]
Приклеюсь к теме recalculate.
Все равно не до конца понимаю логику. 
Вот у меня пример. 
Расчетная таблица в цикле. 

![Screenshot 2022-06-02 at 22 59 41](https://user-images.githubusercontent.com/19141501/171728479-abc5e4f7-8ad6-43df-8132-23063f697c16.png)

В ней одно поле (kolvo_listov) вычисляется по коду в зависимости от левого (kolvo). 
Идея такая. При первоначальном заполнении указывается основная единица измерения в поле кол-во. Она пересчитывается в кол-во листов из справочников типоразмеров. 
Но если пользователь хочет изменить кол-во листов, он может поставить их кол-во в поле kolvo_listov.
В нем стоит код на изменение:

a1=: set(table: $#ntn; field: 'kolvo' = $pesechet_v_osnovn; where: 'id' = #id; log: true)
pesechet_v_osnovn: $#nfv*$pereshet
pereshet: select(table: 'tiporazmery_listov'; field: 'ploshchad'; where: 'id' = #list)

Так вот никакие recalculate  после этого кода, у меня не приводят к пересчету поля кол-во. Только рефреш окна.
Что не так делаю?

**Reply by ttm-alexeypolunin** [reply]
Я думаю там у нас ошибка с обновлением отображения. Если после изменения руками обновить страницу?

**Reply by juramebel** [reply]
Вот так происходит.

https://user-images.githubusercontent.com/19141501/171791358-ca9d8840-565c-4500-b197-5e5fc32351a5.mp4

**Reply by Makhmut2445** [reply]
кол-во основное поставьте сортом 28, после поле которое вы изменяете, и рекалк не нужен будет...

**Reply by Makhmut2445** [reply]
https://user-images.githubusercontent.com/103171127/171798839-787abdd3-e2f7-4332-bbac-534dbdae1ee3.mp4

**Reply by ttm-alexeypolunin** [reply]
**Ошибка в обновлении страницы подтверждается** — если в расчетных таблицах set идет в ту-же таблицу по изменению то изменение полей не отображается. Поправим к понедельнику скорее всего.

**Reply by ttm-alexeypolunin** [reply]
**ПОЧИНИЛИ В 3.4.43.0**
